eos-updater-flatpak-installer/main.c

```diff
@@ -0,0 +1,246 @@
+       const gchar    *error_message,
+       ...) G_GNUC_PRINTF (2, 3);
+
+static int
+usage (GOptionContext *context,
+       const gchar    *error_message,
+       ...)
+{
+  va_list ap;
+  g_autofree gchar *formatted_message = NULL;
+  g_autofree gchar *help = NULL;
+
+  /* Format the arguments. */
+  va_start (ap, error_message);
+  formatted_message = g_strdup_vprintf (error_message, ap);
+  va_end (ap);
+
+  /* Include the usage. */
+  help = g_option_context_get_help (context, TRUE, NULL);
+  g_printerr ("%s: %s\n\n%s\n", g_get_prgname (), formatted_message, help);
+
+  return 128;
+}
+
+static gboolean
+parse_installer_mode (const gchar              *mode,
+                      EosUpdaterInstallerMode  *out_mode,
+                      GError                  **error)
+{
+  GEnumClass *enum_class = g_type_class_ref (EOS_TYPE_UPDATER_INSTALLER_MODE);
+  GEnumValue *enum_value = g_enum_get_value_by_nick (enum_class, mode);
+
+  g_type_class_unref (enum_class);
+
+  if (!enum_value)
+    {
+      g_set_error (error,
+                   G_IO_ERROR,
+                   G_IO_ERROR_FAILED,
+                   "Invalid installer mode %s",
+                   mode);
+      return FALSE;
+    }
+
+  *out_mode = (EosUpdaterInstallerMode) enum_value->value;
+
+  return TRUE;
+}
+
+int
+main (int    argc,
+      char **argv)
+{
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GOptionContext) context = NULL;
+  g_autoptr(GFile) flatpaks_to_export_file = NULL;
+  g_autoptr(GHashTable) flatpak_ref_actions_for_this_boot = NULL;
+  g_autoptr(GHashTable) flatpak_ref_actions_progress = NULL;
+  g_autoptr(FlatpakInstallation) installation = NULL;
+  g_autofree gchar *formatted_flatpak_ref_actions_for_this_boot = NULL;
+  g_autofree gchar *formatted_flatpak_ref_actions_progress_for_this_boot = NULL;
+  const gchar *resolved_mode = NULL;
+  const gchar *pending_flatpak_deployments_state_path = eos_updater_util_pending_flatpak_deployments_state_path ();
+  EosUpdaterInstallerMode parsed_mode;
+
+  gchar *mode = NULL;
+  gboolean also_pull = FALSE;
+  GOptionEntry entries[] =
+  {
+    { "mode", 'm', 0, G_OPTION_ARG_STRING, &mode, "Mode to use (perform, stamp, check)", NULL },
+    { "pull", 'p', 0, G_OPTION_ARG_NONE, &also_pull, "Also pull flatpaks", NULL },
+    { NULL }
+  };
+
+  setlocale (LC_ALL, "");
+
+  context = g_option_context_new ("— Endless OS Updater - Flatpak Installer");
+  g_option_context_add_main_entries (context, entries, NULL);
+  g_option_context_set_summary (context,
+                                "Install flatpak packages on system boot");
+
+  if (!g_option_context_parse (context, &argc, &argv, &error))
+    return usage (context, "Failed to parse options: %s", error->message);
+
+  resolved_mode = mode != NULL ? mode : "perform";
+
+  if (!parse_installer_mode (resolved_mode,
+                             &parsed_mode,
+                             &error))
+    {
+      g_message ("%s", error->message);
+      return EXIT_FAILURE;
```

Comment:
It would make the command line error handling output a little more consistent if you used the `fail()` function from `eos-updater-avahi.c` here (just copy it across, like you’ve done for `usage()`).